photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/AbstractOptimizer.scala

```diff
@@ -44,7 +44,17 @@ abstract class AbstractOptimizer[Datum <: DataPoint, -Function <: DiffFunction[D
     None
   }
 
-  def isDone: Boolean = convergenceReason.nonEmpty
+  def isDone: Boolean = {
+    val isDone = convergenceReason.nonEmpty
+    if (isDone) {
+      // Set the convergenceReason when the optimization is done
+      statesTracker match {
+        case Some(y) => y.convergenceReason = convergenceReason
+        case None =>
+      }
+    }
+    isDone
+  }
```

Comment:
This is semantically a little weird. The name ```isDone``` suggests a simple predicate function with no side-effects, but this change introduces the side-effect of changing the state of ```stateTracker```. Maybe the mutation should be in ```setCurrentState```, but with a check for ```isDone``` as above?